modal/_serialization.py

```diff
@@ -411,75 +412,112 @@ class ParamTypeInfo:
+def deserialize_proto_params(serialized_params: bytes) -> dict[str, Any]:
     proto_struct = api_pb2.ClassParameterSet()
     proto_struct.ParseFromString(serialized_params)
-    value_by_name = {p.name: p for p in proto_struct.parameters}
     python_params = {}
-    for schema_param in schema:
-        if schema_param.name not in value_by_name:
-            # TODO: handle default values? Could just be a flag on the FunctionParameter schema spec,
-            #  allowing it to not be supplied in the FunctionParameterSet?
-            raise AttributeError(f"Constructor arguments don't match declared parameters (missing {schema_param.name})")
-        param_value = value_by_name[schema_param.name]
-        if schema_param.type != param_value.type:
-            raise ValueError(
-                "Constructor arguments types don't match declared parameters "
-                f"({schema_param.name}: type {schema_param.type} != type {param_value.type})"
-            )
+    for param in proto_struct.parameters:
         python_value: Any
-        if schema_param.type == api_pb2.PARAM_TYPE_STRING:
-            python_value = param_value.string_value
-        elif schema_param.type == api_pb2.PARAM_TYPE_INT:
-            python_value = param_value.int_value
-        elif schema_param.type == api_pb2.PARAM_TYPE_BYTES:
-            python_value = param_value.bytes_value
+        if param.type == api_pb2.PARAM_TYPE_STRING:
+            python_value = param.string_value
+        elif param.type == api_pb2.PARAM_TYPE_INT:
+            python_value = param.int_value
+        elif param.type == api_pb2.PARAM_TYPE_BYTES:
+            python_value = param.bytes_value
         else:
-            # TODO(elias): based on `parameters` declared types, we could add support for
-            #  custom non proto types encoded as bytes in the proto, e.g. PARAM_TYPE_PYTHON_PICKLE
-            raise NotImplementedError("Only strings and ints are supported parameter value types at the moment")
+            raise NotImplementedError(f"Unimplemented parameter type: {param.type}.")
 
-        python_params[schema_param.name] = python_value
+        python_params[param.name] = python_value
 
     return python_params
 
 
+def validate_params(params: dict[str, Any], schema: typing.Sequence[api_pb2.ClassParameterSpec]):
+    # first check that all declared values are provided
+    for schema_param in schema:
+        if schema_param.name not in params:
+            # we expect all values to be present - even defaulted ones (defaults are applied on payload construction)
+            raise InvalidError(f"Missing required parameter: {schema_param.name}")
+        python_value = params[schema_param.name]
+        python_type = type(python_value)
+        param_protobuf_type = validate_parameter_type(python_type)
+        if schema_param.type != param_protobuf_type:
+            expected_python_type = PROTO_TYPE_INFO[schema_param.type].type
+            raise TypeError(
```

Comment:
I feel like we need a clear policy about when we use `InvalidError` (or other `modal.exception` types vs. builtin types).

Reply:
Yeah... I didn't think a ton about it here, and went with something like "if the wrong type is provided, use TypeError, if something else doesn't match assumptions, use InvalidError"

Reply:
I don't think you need to hold up this PR on it but we should figure it out as part of 1.0 (there's a ticket already). We're pretty inconsistent about it AFAICT.